### PR TITLE
Refactor disabling hotpatch

### PR DESCRIPTION
**Pull request by panpawn**

### CODE REVIEW

File: chat-commands.js [hunk 1]
```diff
@@ -2468,18 +2469,28 @@ exports.commands = {
 				Chat.uncacheTree('./tournaments');
 				global.Tournaments = require('./tournaments');
 				Tournaments.tournaments = runningTournaments;
-
+				Chat.notify(rooms, `${prefix}${user.name} used /hotpatch ${target}`);
```
Comment (Zarel) [inline chat-commands.js]: Repeating this ten times is not particularly DRY.

File: chat.js [hunk 2]
```diff
@@ -953,6 +953,20 @@ Chat.escapeHTML = function (str) {
 };
 
 /**
+ * Notifies an array of rooms of a message
```
Comment (Zarel) [inline chat.js]: Why was this renamed and moved out of `Rooms`?

File: chat-commands.js (hunk 1 above)
Comment (Zarel) [inline chat-commands.js]: Simple solution is just to not notify if it gets to an `else` condition.

This might also be split into a different commit/pullreq.

File: chat.js (hunk 2 above)
Comment (panpawn) [inline chat.js]: I think it makes more sense under ``Chat``

File: chat-commands.js (hunk 1 above)
Comment (panpawn) [inline chat-commands.js]: I'll just mention this was changed in the commit summary

Comment (panpawn) [inline chat-commands.js]: Done

**Comment by panpawn** [comment]
Updated

File: chat.js (hunk 2 above)
Comment (Zarel) [inline chat.js]: It doesn't; room-level chat interactions belong in `Rooms`.

File: chat-commands.js [hunk 3]
```diff
@@ -2503,47 +2511,106 @@ exports.commands = {
...
-				Monitor.hotpatchLockChat = reason;
-				return this.sendReply("You have disabled hotpatch until the next server restart.");
-			} else if (target.startsWith('disable')) {
-				let reason = target.split(', ')[1];
-				if (!reason) return this.errorReply("Usage: /hotpatch disable, [reason]");
-				Monitor.hotpatchLock = reason;
-				return this.sendReply("You have disabled hotpatch until the next server restart.");
+				this.sendReply("Dnsbl has been hot-patched.");
+			} else {
+				return this.errorReply("Your hot-patch command was unrecognized.");
```
Comment (Zarel) [inline chat-commands.js]: Yes, this is what I was thinking of; this is good. :)

File: chat.js (hunk 2 above)
Comment (Slayer95) [inline chat.js]: +1 to staying in ``Rooms``.

... This part should have been its own commit and even PR too.

**Comment by panpawn** [comment]
Updated

File: chat-commands.js [hunk 4]
```diff
@@ -2468,18 +2469,25 @@ exports.commands = {
...
 			} else if (target === 'battles') {
+				if (hotpatchLock['battles']) return this.errorReply(`Hot-patching battles has been disabled by ${hotpatchLock['battles'].by} (${hotpatchLock['battles'].reason})`);
+
 				Rooms.SimulatorProcess.respawn();
-				return this.sendReply("Battles have been hotpatched. Any battles started after now will use the new code; however, in-progress battles will continue to use the old code.");
+				this.sendReply("Battles have been hot-patched. Any battles started after now will use the new code; however, in-progress battles will continue to use the old code.");
 			} else if (target === 'formats') {
+				if (hotpatchLock['formats']) return this.errorReply(`Hot-patching formats has been disabled by ${hotpatchLock['formats'].by} (${hotpatchLock['formats'].reason})`);
+				if (hotpatchLock['battles']) return this.errorReply(`Hot-patching battles has been disabled by ${hotpatchLock['battles'].by} (${hotpatchLock['battles'].reason})`);
+				if (hotpatchLock['validator']) return this.errorReply(`Hot-patching the validator has been disabled by ${hotpatchLock['validator'].by} (${hotpatchLock['validator'].reason})`);
```
Comment (panpawn) [inline chat-commands.js]: Do these checks look good here? I was semi-confused as to what should disable what, so if this isn't correct, please let me know

File: chat.js (hunk 2 above)
Comment (Zarel) [inline chat.js]: You should probably have commented in #3361, @Slayer95 :|

File: chat-commands.js (hunk 4 above)
Comment (Zarel) [inline chat-commands.js]: Yeah, that's good, although `hotpatchLock['formats']` is kind of unnecessary?

Comment (Zarel) [inline chat-commands.js]: I guess it makes a better error message; it can stay.

Comment (Zarel) [inline chat-commands.js]: You should check for `hotpatchLock['formats']` in battle and validator hotpatches, though.

File: chat-commands.js [hunk 5]
```diff
@@ -2503,47 +2511,102 @@ exports.commands = {
...
 			} else if (target === 'dnsbl' || target === 'datacenters') {
 				Dnsbl.loadDatacenters();
-				return this.sendReply("Dnsbl has been hotpatched.");
-			} else if (target.startsWith('disablechat')) {
-				if (Monitor.hotpatchLockChat) return this.errorReply("Hotpatch is already disabled.");
-				let reason = target.split(', ')[1];
-				if (!reason) return this.errorReply("Usage: /hotpatch disablechat, [reason]");
-				Monitor.hotpatchLockChat = reason;
-				return this.sendReply("You have disabled hotpatch until the next server restart.");
-			} else if (target.startsWith('disable')) {
```
Comment (Zarel) [inline chat-commands.js]: We might as well put a message about `/nohotpatch`

Comment (panpawn) [inline chat-commands.js]: done

File: chat-commands.js (hunk 4 above)
Comment (panpawn) [inline chat-commands.js]: done

**Comment by panpawn** [comment]
Updated

File: chat-commands.js [hunk 6]
```diff
@@ -2443,14 +2443,15 @@ exports.commands = {
 
 	hotpatch: function (target, room, user) {
 		if (!target) return this.parse('/help hotpatch');
-		if (!this.can('hotpatch')) return false;
-		if (Monitor.hotpatchLock) return this.errorReply("Hotpatch is currently been disabled. (" + Monitor.hotpatchLock + ")");
-
-		Rooms.global.notifyRooms(['development', 'staff', 'upperstaff'], `|c|${user.getIdentity()}|/log ${user.name} used /hotpatch ${target}`);
+		if (!this.can('hotpatch')) return;
+		const rooms = ['development', 'staff', 'upperstaff'];
+		const prefix = `|c|${user.getIdentity()}|/log `;
```
Comment (Zarel) [inline chat-commands.js]: `notifyRooms` is only called once, so all of this can be moved down next to it.

File: chat-commands.js [hunk 7]
```diff
@@ -2443,14 +2443,15 @@ exports.commands = {
...
+		const hotpatchLock = Monitor.hotpatchLock;
```
Comment (Zarel) [inline chat-commands.js]: Defining a variable seems overkill to chop off literally `Monitor.` - at least call it `lock`.

File: chat-commands.js (hunk 6 above)
Comment (panpawn) [inline chat-commands.js]: It's called twice, but I can still move it down

Comment (panpawn) [inline chat-commands.js]: Moved it down

File: chat-commands.js (hunk 7 above)
Comment (panpawn) [inline chat-commands.js]: I renamed it to ``lock``

File: chat-commands.js [hunk 8]
```diff
@@ -2503,47 +2512,106 @@ exports.commands = {
...
-				if (!reason) return this.errorReply("Usage: /hotpatch disablechat, [reason]");
-				Monitor.hotpatchLockChat = reason;
-				return this.sendReply("You have disabled hotpatch until the next server restart.");
+				this.sendReply("Dnsbl has been hot-patched.");
 			} else if (target.startsWith('disable')) {
-				let reason = target.split(', ')[1];
-				if (!reason) return this.errorReply("Usage: /hotpatch disable, [reason]");
-				Monitor.hotpatchLock = reason;
-				return this.sendReply("You have disabled hotpatch until the next server restart.");
+				this.sendReply("Disabling hot-patch has been moved to its' own command:");
```
Comment (Zarel) [inline chat-commands.js]: No `'` at all.

Comment (panpawn) [inline chat-commands.js]: Done

(on a side-note, I really need to practice when and where to use apostrophes, lol)

File: chat-commands.js [hunk 9]
```diff
@@ -2503,47 +2511,105 @@ exports.commands = {
...
+				by: user.name,
+				reason: reason,
+			};
+			this.sendReply("You have disabled hot-patching tournaments.");
+		} else {
+			return this.errorReply("This hotpatch is not an option to disable.");
+		}
+		Rooms.global.notifyRooms(['development', 'staff', 'upperstaff'], `|c|${user.getIdentity()}|/log ${user.name} has disabled hot-patching ${hotpatch}.`);
+	},
+	disablehotpatchhelp: ["/disablehotpatch [chat|formats|battles|validator|tournaments] [reason] - Disables hotpatching the specified part of the simulator. Requires: ~"],
```
Comment (panpawn) [inline chat-commands.js]: Should there be a 'general' option to disable all hotpatching? Is that functionality worth keeping?

**Comment by panpawn** [comment]
Note: These changes **can** be hotpatched into the production server, but if they are, ``Monitor.hotpatchLock`` needs to be set to ``{}``

File: chat-commands.js (hunk 9 above)
Comment (Zarel) [inline chat-commands.js]: There's practically no reason to disable all hotpatching simultaneously, afaik. Like, if it were ever used, it'd be as a convenience thing, and that's just a dick move.

I _guess_ it'd be used for major refactors (renaming a global variable?) but still, I don't want people to use it out of laziness. They can nohotpatch chat, formats, and tours separately.

Comment (panpawn) [inline chat-commands.js]: Okay, just wanted to make sure

Comment (Zarel) [inline chat-commands.js]: One thing we can do is support `all`, but disable it on Main.

On other, smaller servers, being lazy doesn't have as many severe consequences.

Comment (Zarel) [inline chat-commands.js]: By which I mean, support both `/hotpatch all` and `/nohotpatch all`; we used to have `/hotpatch all` but removed it because it did too much damage to Main.

Comment (panpawn) [inline chat-commands.js]: ok

Comment (panpawn) [inline chat-commands.js]: Oh, I completely missed your second comment...

**Comment by panpawn** [comment]
So, I simplified ``/disablehotpatch`` and I also supported ``/hotpatch all`` as an option to enable in ``Config``.

Does this look good?

File: chat-commands.js [hunk 10]
```diff
@@ -2503,47 +2519,81 @@ exports.commands = {
...
-		"/hotpatch disable, [reason] - disables the use of hotpatch until the next server restart"],
+		"/hotpatch punishments - reloads new punishments code",
+		"/hotpatch tournaments - reloads new tournaments code",
+		"/hotpatch all - hot-patches chat, tournaments, formats, login server, punishments, and dnsbl",
+	],
+
+	hotpatchlock: 'disablehotpatch',
+	nohotpatch: 'disablehotpatch',
+	hotpatchdisable: 'disablehotpatch',
+	disablehotpatch: function (target, room, user) {
```
Comment (Slayer95) [inline chat-commands.js]: Do we really need 4 versions of a command which is only expected to be used very sparingly by a handful of people?

Comment (panpawn) [inline chat-commands.js]: I mean, the entire point of an alias is if to make sure someone can use the command if they don't know the exact command name, I'd say if anything, that this is a pretty good usage of aliases.

It's not like other commands would/should be using any of these aliases, anyways

Comment (Zarel) [inline chat-commands.js]: I also think `nohotpatch` should be default, the others are too long.

Comment (panpawn) [inline chat-commands.js]: I made ``nohotpatch`` the default

File: chat-commands.js [hunk 11]
```diff
@@ -2503,47 +2519,81 @@ exports.commands = {
...
-		"/hotpatch disable, [reason] - disables the use of hotpatch until the next server restart"],
+		"/hotpatch punishments - reloads new punishments code",
+		"/hotpatch tournaments - reloads new tournaments code",
+		"/hotpatch all - hot-patches chat, tournaments, formats, login server, punishments, and dnsbl",
+	],
+
+	hotpatchlock: 'nohotpatch',
+	disablehotpatch: 'nohotpatch',
+	hotpatchdisable: 'nohotpatch',
+	nohotpatch: function (target, room, user) {
```
Comment (Slayer95) [inline chat-commands.js]: IMO aliases should only be used for user-facing commands.

This is the total opposite, a dev-only command, and we can probably expect devs to be able to figure out what the right command is, either by reading the code (if you think that disabling hotpatch is needed, you are likely to be already in Github) or by reading the manual (mention this feature in the ``/help hotpatch``)?

IMO we should just keep ``/hotpatchlock`` and ``/nohotpatch``

Comment (panpawn) [inline chat-commands.js]: I mean, ``/help blacklist`` doesn't mention the other blacklist-related commands... It seems weird to me to document a dev command more than a much more commonly used command.

If you feel really strongly about not using so many aliases, then I can remove some - I just don't see why it's such a big deal myself

**Comment by panpawn** [comment]
I updated this to mention ``/nohotpatch`` in ``/help hotpatch`` and I removed the other aliases

**Comment by Slayer95** [comment]
Thanks!

Tangential:
> I mean, /help blacklist doesn't mention the other blacklist-related commands...

I'd definitely support enhancing the documentation of namespace commands like that. (...Because ``blacklist`` is namespace, right?)

**Comment by panpawn** [comment]
>Because blacklist is namespace, right?)

Nope

File: chat-commands.js (hunk 10 above)
Comment (Zarel) [inline chat-commands.js]: I'd generally agree with panpawn re: "ability to guess the command name"

File: chat-commands.js (hunk 11 above)
Comment (Zarel) [inline chat-commands.js]: I'm leaning towards "let the aliases stay", but not strongly. These names shouldn't be used for any other command, and letting a dev guess a command name is mildly useful for the same reasons any other alias is useful.

On the other hand, if @Slayer95 _really_ cares, I don't care enough to overrule him. He can decide, I guess.

File: chat-commands.js (hunk 10 above)
Comment (panpawn) [inline chat-commands.js]: ... should I add the aliases back?

File: chat-commands.js (hunk 11 above)
Comment (panpawn) [inline chat-commands.js]: Yeah, I don't feel strongly either way - I just don't see how this being a dev command should have to do with a few less aliases is all

File: config/config-example.js [hunk 12]
```diff
@@ -228,6 +228,11 @@ exports.repl = true;
 exports.replsocketprefix = './logs/repl/';
 exports.replsocketmode = 0o600;
 
+// allowhotpatchall - allows for hotpatching every hotpatchable part of PS
+// at once. Generally speaking, there probably won't be a need to for this,
+// which is why this is disabled by default.
+exports.allowhotpatchall = false;
```
Comment (Zarel) [inline config/config-example.js]: Should be `true` by default; Main is the only server where this should be off.

Or, well, it should be named `disablehotpatchall = false` by default.

Comment (panpawn) [inline config/config-example.js]: done

**Comment by panpawn** [comment]
Updated

File: config/config-example.js [hunk 13]
```diff
@@ -228,6 +228,11 @@ exports.repl = true;
 exports.replsocketprefix = './logs/repl/';
 exports.replsocketmode = 0o600;
 
+// disablehotpatchall - disables allowing hotpatching every hotpatchable part
+// of PS at once. Generally speaking, there probably won't be a need to for this,
+// which is why there's an option to disable this.
+exports.disablehotpatchall = false;
```
Comment (Zarel) [inline config/config-example.js]: I wonder if we should have it in `config-example` at all? I think most servers will want to be able to `/hotpatch all`.

Comment (panpawn) [inline config/config-example.js]: I will admit, I'm a fan of documenting ``Config`` settings.  I don't really see any harm for it in this case.

File: config/config-example.js [hunk 14]
```diff
@@ -228,6 +228,11 @@ exports.repl = true;
 exports.replsocketprefix = './logs/repl/';
 exports.replsocketmode = 0o600;
 
+// disablehotpatchall - disables allowing hotpatching every hotpatchable part
+// of PS at once. Generally speaking, there probably won't be a need to for this,
+// which is why there's an option to disable this.
```
Comment (Zarel) [inline config/config-example.js]: Generally speaking, there's a pretty big need for `/hotpatch all` - convenience. The only advantage any hotpatch forms other than `all` is lower RAM use (which is only a problem for Main because Main is huge), and to do pinpoint hotpatching (like `/disablehotpatch`).

Comment (Zarel) [inline config/config-example.js]: (I'm asking you to change this description)

Comment (Zarel) [inline config/config-example.js]: > `disables allowing`

Just ``disables `/hotpatch all` ``

File: config/config-example.js (hunk 13 above)
Comment (Zarel) [inline config/config-example.js]: Fair. Leave it in, I can understand how someone might want this.

File: config/config-example.js (hunk 14 above)
Comment (panpawn) [inline config/config-example.js]: >(I'm asking you to change this description)

honestly didn't realize that until i saw that comment and I was really confused, lol

**Comment by panpawn** [comment]
Updated Config documentation comment

**Comment by Zarel** [comment]
Looks great!
